### DISCUSSION TITLE
Does Flix support Rank-2 Types?

### DISCUSSION THREAD

**Discussion by thedeemon** [flix/flix#6258] [General]
Trying out version 0.38, type checker fails to unify a type variable with Int32:
```
def hey(f : a -> (a,a)) : Unit \ IO =
    f(1) |> println
```
saying
```
>> Expected type: 'a' but found type: 'Int32'.
8 |     f(1) |> println
          ^
          expression has unexpected type.
```
Adding `[a : Type]` doesn't change it.

I didn't expect that from a HM-based system. Am I doing something wrong or is it a bug?

**Comment by magnus-madsen** [comment]
If `f` has type `a -> (a, a)` then it can only be applied to elements of type `a`. You cannot apply it to e.g. `Int32`. 

This is not a bug, but how it is supposed to work.

If you want to apply it to an `Int32` then it needs to have type `Int32 -> (Int32, Int32)`.

**Reply by magnus-madsen** [reply]
Or to think of it differently, if you have the expression `f(1)` then clearly `f` must be prepared to accept an `Int32`.

On the other hand, if you have `f(x)` then all you need is that `x` and the argument of `f` should have the same type, so here polymorphism may be possible.

**Reply by thedeemon** [reply]
I can define a polymorphic function
```
def dbl(x : a) : (a,a) = (x,x)
```
whose type is `a -> (a,a)`. It can take any type on input:
```
def hoy() : (Int32, String) =
    let a = dbl(1);
    let b = dbl("aa");
    (fst(a), snd(b))
```
This works fine.

Here Flix is able to instantiate the type scheme twice and unify `a` with different input types.

Now, can I pass this polymorhic function `dbl` as an argument to another function? Instead of using it directly as in `hoy()` I want to call it as in `hey(f)` above. Hence my original question.

**Reply by magnus-madsen** [reply]
Are you familiar with rank-2 types? It seems that is what you are asking for. (Or maybe there is just confusion?)

(Flix - as most HM languages - does not have higher-ranked types).

**Reply by thedeemon** [reply]
Right, this would be a rank-2 type. IIRC both OCaml and Haskell have them. 

I've read Flix uses monomorphisation and I'm curious whether polymorphic functions can be first class values there.

**Reply by magnus-madsen** [reply]
I believe it is an open research question how to combine monomorphization and rank-2 types. 

There is a rumor that the group around Effekt lang is working on it.

**Reply by thedeemon** [reply]
Got it, thanks!

**Comment by magnus-madsen** [comment]
TLDR: Question was whether Flix has rank-2 types-- which is currently not the case.
